File: python/sdmd_run.py
import dmdtools
import numpy as np
import matplotlib.pyplot as plt

max_rank = 10                   # maximum allowable rank of the DMD operator
                                #   (set to zero for unlimited)
n_snaps = 501                   # total number of snapshots to be processed
n_states = 4000                 # number of states
noise_cov = 1.e-4               # measurement noise covariance

dt = 0.01                       # timestep

np.random.seed(0)

def snapshots(n_states, n_snaps, noise_cov=0):
    # Define the example system
    v1 = np.random.randn(n_states)
    v2 = np.random.randn(n_states)
    v3 = np.random.randn(n_states)
    v4 = np.random.randn(n_states)
    # characteristic frequencies
    f1 = 5.2
    f2 = 1.0
    for k in range(n_snaps):
        x = (v1 * np.cos(2 * np.pi * f1 * dt * k) +
             v2 * np.cos(2 * np.pi * f2 * dt * k) +
             v3 * np.sin(2 * np.pi * f1 * dt * k) +
             v4 * np.sin(2 * np.pi * f2 * dt * k))
        yield x + np.sqrt(noise_cov) * np.random.randn(n_states)

def standard_dmd():
    X = np.zeros((n_states, n_snaps-1))
    Y = np.zeros((n_states, n_snaps-1))
    snaps = snapshots(n_states, n_snaps, noise_cov)
    x = snaps.next()
    for k, y in enumerate(snaps):
        X[:,k] = x
        Y[:,k] = y
        x = y
    return dmdtools.dmd(X, Y)

def streaming_dmd():
    sdmd = dmdtools.StreamingDMD(max_rank)
    snaps = snapshots(n_states, n_snaps, noise_cov)
    x = snaps.next()
    for y in snaps:
        sdmd.update(x, y)
        x = y
    return sdmd.compute_modes()

def main(streaming):
    modes, evals = streaming_dmd() if streaming else standard_dmd()
    fdmd = np.abs(np.angle(evals)) / (2 * np.pi * dt)
    n_modes = len(fdmd)
    ydmd = np.zeros(n_modes)
    for i in range(n_modes):
        ydmd[i] = np.linalg.norm(modes[:,i] * np.abs(evals[i]))
    ydmd /= max(ydmd)
    plt.stem(fdmd, ydmd)
    plt.show()

def compare_methods():
    np.random.seed(0)
    modes, evals = standard_dmd()

    np.random.seed(0)
    modes2, evals2 = streaming_dmd()

    evals.sort()
    evals2.sort()
    # print("standard:")
    # print(evals)
    # print("\nstreaming:")
    # print(evals2)
    print(np.allclose(evals, evals2))

if __name__ == "__main__":
    streaming = True
    main(streaming)
    # compare_methods()

File: matlab/sdmd_run.m
%SDMD_RUN An example to demonstrate incrementally updated DMD.
%
%   Run a simple example of incrementally updated DMD on a toy problem to
%   demonstrate the streaming DMD algorithm, as formulated in:
%   
%       M.S. Hemati, M.O. Williams, C.W. Rowley, 
%       ``Dynamic Mode Decomposition for Large and Streaming Datasets,'' 
%       Physics of Fluids (2014).
%
%   The example considers an arbitrary n-dimensional dynamical system with 
%   two characteristic frequencies.
%
%   A total of m snapshot pairs are measured sequentially and are subject 
%   to additive i.i.d. zero-mean Gaussian noise with covariance noise_cov.
%
%   To toggle between streaming and batch-processed DMD modes, toggle the
%   streaming flag between true and false, respectively.  The StreamingDMD 
%   Matlab class is used to perform incremental updates to the DMD 
%   computation when streaming=true.
%
%   When in streaming mode, to run the direct streaming algorithm set 
%   max_rank=0; otherwise, incrementally update DMD with POD truncation,
%   where max_rank denotes the truncation threshold (i.e., the maximum
%   number of retained POD modes).
%
% Tested using Matlab 8.1.0.604 (R2013a) for Mac OSX.
%
% Authors:
%   Maziar S. Hemati
%   Matthew O. Williams
%   Clarence W. Rowley
%
% Created:
%   August 2014.
%
%   see also StreamingDMD
%
% Reference page in Help browser:
%   <a href="matlab:doc sdmd_run">doc sdmd_run</a>



% set example parameters here
max_rank = 10;      % maximum allowable rank of the DMD operator
                    %   (set to zero for unlimited)
m = 501;            % total number of snapshots to be processed
n = 4000;           % number of states
noise_cov = 1e-4;   % measurement noise covariance
rng(0)              % seed the random number generator
streaming = true;   % true=use streaming DMD, false=use batch-processed DMD

%% define the example system
% random state directions
v1 = randn(n,1);
v2 = randn(n,1);
v3 = randn(n,1);
v4 = randn(n,1);

% characteristic frequencies
f1 = 5.2;
f2 = 1.0;

% sampling time
dt = 1e-2;

dynsys = @(k) v1*cos(2*pi*dt*f1*k) + v2*cos(2*pi*f2*dt*k) + ...
    v3*sin(2*pi*dt*f1*k) + v4*sin(2*pi*f2*dt*k);

get_snapshot = @(k) dynsys(k) + sqrt(noise_cov) * randn(n,1);

%% collect the data

disp('Collecting data')
tic
if ~streaming
    % standard algorithm: batch-processing
    X = zeros(n, m);
    Y = zeros(n, m);
    yk = get_snapshot(0);
    for k = 1:m
        xk = yk;
        yk = get_snapshot(k);
        X(:,k) = xk;
        Y(:,k) = yk;
    end
    [Qx, S, V] = svd(X, 0);
    Ktilde = Qx' * Y * V * pinv(S);
else
    % streaming algorithm
    sdmd = StreamingDMD(max_rank);
    yk = get_snapshot(0);
    for k = 1:m
        xk = yk;
        yk = get_snapshot(k);
        sdmd = sdmd.update(xk, yk);
    end
end
elapsed_time = toc;
fprintf('  Elapsed time: %f seconds\n', elapsed_time)


%% compute DMD spectrum
disp('Computing spectrum')
tic
if ~streaming
    % standard DMD spectrum
    [evecK, evals] = eig(Ktilde);
    evals = diag(evals);
    modes = Qx * evecK;
else
    % streaming algorithm
    [modes, evals] = sdmd.compute_modes();
end
elapsed_time = toc;
fprintf('  Elapsed time: %f seconds\n', elapsed_time)

% calculate corresponding frequencies
fdmd = abs(angle(evals)) ./ (2 * pi * dt);
ydmd = zeros(length(fdmd),1);
for ii = 1:length(fdmd)
    ydmd(ii) = norm(modes(:,ii)).*abs(evals(ii));
end
ydmd = ydmd./max(ydmd);

figure(1)
stem(fdmd,ydmd,'o-')
xlabel('Frequency')
ylabel('Magnitude')
